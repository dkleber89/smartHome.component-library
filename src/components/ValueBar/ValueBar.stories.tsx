import React, { ReactElement } from 'react';
import { withKnobs, number } from '@storybook/addon-knobs';

import ValueBar from './ValueBar';

export default {
  title: 'ValueBar',
  component: ValueBar,
  decorators: [withKnobs],
  parameters: {
    info: ``,
  },
};

export const DefaultValues = (): ReactElement => <ValueBar value={number('Value', 80, { min: 0, max: 100 })} />;
export const CustomValues = (): ReactElement => <ValueBar min={-500} max={500} value={250} unit="mm" />;
export const ProgressIndeterminate = (): ReactElement => <ValueBar />;
