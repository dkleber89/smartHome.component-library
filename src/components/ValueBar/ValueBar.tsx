import React, { ReactElement, useCallback } from 'react';
import { LinearProgress as MuiLinearProgress, Box as MuiBox, Typography as MuiTypography } from '@material-ui/core';
import styled from 'styled-components';

export interface IValueBarProps {
  /** Component className */
  className?: string;
  /** Actual Value */
  value?: number;
  /** Max Value */
  max?: number;
  /** Min Value */
  min?: number;
  /** Scale Unit */
  unit?: string;
  /** Primary or Secondary Color */
  color?: 'primary' | 'secondary';
}

const StyledValueBar = styled.div<IValueBarProps>`
  width: 100%;
`;

const StyledLinearProgress = styled(MuiLinearProgress)`
  height: 10px;
`;

const ValueBar = ({ className, value, min = 0, max = 100, unit = '%', color = 'primary' }: IValueBarProps): ReactElement => {
  const normalise = useCallback(
    (progressValue: number | undefined): number | undefined => {
      if (typeof progressValue === 'number') {
        return ((progressValue - min) * 100) / (max - min);
      }

      return undefined;
    },
    [min, max]
  );

  return (
    <StyledValueBar className={className}>
      <MuiBox display="flex" alignItems="center">
        <MuiBox width="100%" mr={1}>
          <StyledLinearProgress color={color} variant={value ? 'determinate' : 'indeterminate'} value={normalise(value)} />
        </MuiBox>
        {value && (
          <MuiBox minWidth={35} data-testid="value-text">
            <MuiTypography variant="body2" color="textSecondary">{`${Math.round(value)}${unit}`}</MuiTypography>
          </MuiBox>
        )}
      </MuiBox>
    </StyledValueBar>
  );
};

export default ValueBar;
