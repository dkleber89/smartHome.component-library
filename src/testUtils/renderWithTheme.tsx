import React, { ReactElement } from 'react';
import { render, RenderResult, RenderOptions } from '@testing-library/react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import { StylesProvider, ThemeProvider as MuiThemeProvider, Theme } from '@material-ui/core';

import ITheme, { lightTheme } from '../themes';

function renderWithTheme(component: ReactElement, options?: RenderOptions, theme: Theme & ITheme = lightTheme): RenderResult {
  const Providers = (): ReactElement => {
    return (
      <StylesProvider injectFirst>
        <StyledThemeProvider theme={theme}>
          <MuiThemeProvider theme={theme}>{component}</MuiThemeProvider>
        </StyledThemeProvider>
      </StylesProvider>
    );
  };

  const rendered = render(component, { wrapper: Providers, ...options });
  return {
    ...rendered,
    rerender: (ui: ReactElement) => renderWithTheme(ui, { container: rendered.container, ...options }, theme),
  };
}

export default renderWithTheme;
