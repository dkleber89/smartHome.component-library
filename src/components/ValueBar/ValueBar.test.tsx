import React from 'react';
import { RenderResult } from '@testing-library/react';
import renderWithTheme from '../../testUtils';

import ValueBar from './ValueBar';

describe('ValueBar test', () => {
  test('snapshot', () => {
    const { container } = renderWithTheme(<ValueBar />);

    expect(container.firstChild).toMatchSnapshot();
  });

  // Test Props -> Snippet: testProps
  describe('props', () => {
    let renderResult: RenderResult;

    beforeEach(() => {
      renderResult = renderWithTheme(<ValueBar value={80} />);
    });

    test('determinated', () => {
      const { rerender, queryByRole, queryByText } = renderResult;

      expect(queryByText('80%')).toBeInTheDocument();
      expect(queryByRole('progressbar')).toHaveAttribute('aria-valuenow', '80');

      rerender(<ValueBar value={300} min={200} max={400} unit="°C" />);

      expect(queryByText('300°C')).toBeInTheDocument();
      expect(queryByRole('progressbar')).toHaveAttribute('aria-valuenow', '50');
    });

    test('indeterminated', () => {
      const { rerender, queryByTestId } = renderResult;

      expect(queryByTestId('value-text')).toBeInTheDocument();

      rerender(<ValueBar />);

      expect(queryByTestId('value-text')).not.toBeInTheDocument();
    });
  });

  // Test Events -> Snippet: testEvents
});
