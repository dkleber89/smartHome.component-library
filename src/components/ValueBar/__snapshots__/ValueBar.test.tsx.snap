// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueBar test snapshot 1`] = `
.c0 {
  width: 100%;
}

.c1 {
  height: 10px;
}

<div
  class="c0"
>
  <div
    class="MuiBox-root MuiBox-root-2"
  >
    <div
      class="MuiBox-root MuiBox-root-3"
    >
      <div
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary c1 MuiLinearProgress-indeterminate"
        role="progressbar"
      >
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
        />
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
        />
      </div>
    </div>
  </div>
</div>
`;
